api/pkg/handler/v1/serviceaccount/sa.go

```
@@ -0,0 +1,121 @@
+		req := request.(AddReq)
+		userInfo := ctx.Value(middleware.UserInfoContextKey).(*provider.UserInfo)
+		err := req.Validate()
+		if err != nil {
+			return nil, err
+		}
+		saFromRequest := req.Body
+		project, err := projectProvider.Get(userInfo, req.ProjectID, &provider.ProjectGetOptions{})
+		if err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+
+		// check if service account name is already reserved in the project
+		existingSA, err := serviceAccountProvider.GetServiceAccountByNameForProject(userInfo, saFromRequest.Name, project.Name)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return nil, common.KubernetesErrorToHTTPError(err)
+			}
+		}
+
+		if existingSA != nil {
+			return nil, common.KubernetesErrorToHTTPError(fmt.Errorf("the given name: '%s' for service account already exists", saFromRequest.Name))
+		}
+
+		groupName := rbac.GenerateActualGroupNameFor(project.Name, saFromRequest.Group)
+		sa, err := serviceAccountProvider.CreateServiceAccount(userInfo, project, saFromRequest.Name, groupName)
+		if err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+
+		return convertInternalServiceAccountToExternal(sa), nil
+	}
+}
+
+// AddReq defines HTTP request for addServiceAccountToProject
+// swagger:parameters addServiceAccountToProject
+type AddReq struct {
+	common.ProjectReq
+	// in: body
+	Body apiv1.ServiceAccount
+}
+
+// Validate validates AddReq request
+func (r AddReq) Validate() error {
+	if len(r.ProjectID) == 0 || len(r.Body.Name) == 0 || len(r.Body.Group) == 0 {
+		return k8cerrors.NewBadRequest("the name, project ID and group cannot be empty")
+	}
+
+	isRequestedGroupPrefixValid := false
+	for _, existingGroupPrefix := range serviceAccountGroupsPrefixes {
+		if existingGroupPrefix == r.Body.Group {
+			isRequestedGroupPrefixValid = true
+			break
+		}
+	}
+	if !isRequestedGroupPrefixValid {
+		return k8cerrors.NewBadRequest("invalid group name %s", r.Body.Group)
```

if the validation is done in the decoding function then the users gets `BadRequest` since this function will not be run there could you return a normal error (`fmt.Errorf`) and wrap it in the endpoint ? This improves readability

---

done